File: frontend/src/app/services/SpotifyDto.ts

interface StringReturn {
    data: string
}
interface CallbackData {
    jwt: string,
    code:  string,
    state: string 
}
interface streamerJsonData {
    name: string,
    path: string
  }
interface TransferPlaylistObject {
    identifier: string,
    data: any
}
interface PlaylistsArray {
    href?: string,
    limit?: number,
    next?: string,
    offset?: number,
    previous?: string,
    total?: number,
    items :itemsOfPlaylist[]
}
interface itemsOfPlaylist {
    collaborative?: boolean, //!-----------------------------
    description?: string, 
    external_urls?: {
        spotify?: string
    },
    href: string,
    id: string,
    images: {
        url: string,
        height: number,
        width: number
    }[],
    name: string,
    owner: {
        external_urls: {
            spotify: string
        },
        followers?: {        //!-----------------------------
            href: string,
            total: number
        },
        href: string,
        id: string,
        type: string,
        uri: string,
        display_name: string
    },
    public: boolean,
    snapshot_id: string,
    tracks?: {
        href: string,
        total: number
    },
    type: string,
    uri: string
}
interface NewTrackToPlaylist {
    track: NewTrack,
    playlistId: string
}
interface NewTrack {
    name: string,
    author: string,
    uri: string,
}
interface NewPlaylist {
    name: string,
    description: string,
    public: boolean
}

interface Playlist {
    next?: string,
    href?: string,
    id?: string,
    name?: string,
    total?: number,
    items :Track[]
}
interface Track {
    added_at?: Date,
    added_by?: UserInfo,
    is_local?: boolean,
    primary_color?: string,
    track: TrackInfo,
    video_thumbnail: {url?: string}
}
interface UserInfo {
    external_urls: {spotify: string},
    href:string,
    id:string,
    type:string,
    uri:string
}
interface TrackInfo {
    album: { 
        album_type: 'album', 
        artists: {name: string}[], 
        available_markets: string[], 
        external_urls: {spotify:string}[],
        images:{url:string}[], 
        href: string
    },
    artists: { name:string }[],
    available_markets: string[],
    disc_number: number,
    duration_ms: number,
    episode: boolean,
    explicit: boolean,
    external_ids: { isrc: string },
    external_urls:{ spotify: string },
    href: string,
    id: string, 
    is_local: boolean,
    name: string,
    popularity: number,
    preview_url: string,
    track: boolean,
    track_number: number,
    type: string,
    uri: string
}

export { StringReturn, CallbackData, PlaylistsArray, itemsOfPlaylist, 
    NewTrack, NewPlaylist, Playlist, Track, TrackInfo, UserInfo, 
    TransferPlaylistObject, NewTrackToPlaylist, streamerJsonData }

File: backend/DTO/SpotifyDto.cs

using music_api.DTO;

namespace music_api.DTO;




public class SpotifyToken
{
    public string access_token { get; set; }
    public string token_type { get; set; }
    public string scope  { get; set; }
    public int expires_in { get; set; }
    public string refresh_token { get; set; }
}


public class CallbackData
{
    public string jwt { get; set; }
    public string code { get; set; }
    public string state { get; set; }
}

public class SpotifyRequesAccessTokenBody
{
    public string GrantType { get; set; }
    public string Code { get; set; }
    public string RedirectUri { get; set; }
}

public class SpotifyRequestRefreshTokenBody {
    public string grant_type { get; set; }
    public string refresh_token { get; set; }
}

//TODO: convert json object in class

public class SpotifyUserData {
  public string country { get; set; }
  public string display_name { get; set; }
  public string email { get; set; }
  public explicit_content explicit_Content { get; set; }
  public external_urls external_urls { get; set; }
  public followers followers { get; set; }
  public string href { get; set; }
  public string id { get; set; }
  public Images[] images { get; set; }
  public string product { get; set; }
  public string type { get; set; }
  public string uri { get; set; }
}
public class explicit_content {
  public bool filter_enabled { get; set; }
  public bool filter_locked { get; set; }
}
public class external_urls {
  public string spotify { get; set; }
}
public class followers {
  public string href { get; set; }
  public int total { get; set; }
}



public class SpotifyUserPlaylists
{
    public string href { get; set; }
    public int limit { get; set; }
    public string next { get; set; }
    public int offset { get; set; }
    public string previous { get; set; }
    public int total { get; set; }
    public SpotifyUserPlaylistItems[] items { get; set; }
}
public class SpotifyUserPlaylistItems
{
  public bool collaborative { get; set; }
  public string description { get; set; }
  public external_urls external_urls { get; set; }
  public string href { get; set; }
  public string id { get; set; }
  public Images[] images { get; set; }
  public string name { get; set; }
  public SpotifyUserPlaylistOwner owner { get; set; }
  public bool @public { get; set; }
  public string snapshot_id { get; set; }
  public HrefTracks tracks { get; set; }
  public string type { get; set; }
  public string uri { get; set; }
}
public class SpotifyUserPlaylistOwner
{
  public external_urls external_urls { get; set; }
  public followers followers { get; set; }
  public string href { get; set; }
  public string id { get; set; }
  public string type { get; set; }
  public string uri { get; set; }
  public string display_name { get; set; }
}

public record CreatePlaylistData
{
	public string name { get; set; }
	public string description { get; set; }
	public bool @public { get; set; }
}
public record UserCreatePlaylistWithJwt
{
  public JWT Jwt { get; set; }
  public CreatePlaylistData Data { get; set; }
}
public record musicData
{
  public string name { get; set; }
  public string author { get; set; }
  public string uri { get; set; }
}

public record TrackAndPlaylist
{
  public musicData Track { get; set; }
  public string PlaylistId { get; set; }
}

public record artist
{
	public external_urls external_urls { get; set; }
	public string href { get; set; }
	public string id { get; set; }
	public string name { get; set; }
}
public record album
{
	public string name { get; set; }
}


public record TrackSearchItems
{
	public album album { get; set; }
	public artist[] artists { get; set; }
	public external_urls external_urls { get; set; }
	public string href { get; set; }
	public string id { get; set; }
	public string name { get; set; }
  public string uri { get; set; }
}
public record Tracks
{
	public string href { get; set; }
	public int limit { get; set; }
	public int offset { get; set; }
	public string next { get; set; }
	public string previous { get; set; }
	public int total { get; set; }
	public TrackSearchItems[] items { get; set; }
}
public record TrackSearchData
{
	public Tracks tracks { get; set; }
}
